module/zfs/metaslab.c

```diff
@@ -2259,6 +2464,34 @@ metaslab_sync(metaslab_t *msp, uint64_t txg)
 	metaslab_class_histogram_verify(mg->mg_class);
 	metaslab_group_histogram_remove(mg, msp);
 
+	/*
+	 * Synchronize the allocation-by-category data into space map
+	 * header, which will be written as part of space_map_write(),
+	 * and reset counters for the next txg sync pass.
+	 */
+	if (msp->ms_category_enabled_birth != 0) {
+		/* TBD -- need an API to activate and check birth time */
+		if (msp->ms_sm->sm_phys->smp_alloc_info.enabled_birth == 0) {
+			msp->ms_sm->sm_phys->smp_alloc_info.enabled_birth =
+			    msp->ms_category_enabled_birth;
+		}
+		space_map_sync_block_allocations(msp->ms_sm,
```

Comment:
Just asking for my own learning process, based on other comments is `dmu_buf_will_dirty(msp->ms_sm->sm_dbuf, tx);` the right thing to add right around here? Fixes problems in `space_map.c`